pytorch_lightning/core/lightning.py

```diff
@@ -1998,3 +2001,17 @@ def get_from_queue(self, queue: torch.multiprocessing.SimpleQueue) -> None:
         self.trainer.callback_metrics.update(
             apply_to_collection(callback_metrics, np.ndarray, lambda x: torch.tensor(x))
         )
+
+    @contextmanager
+    def prevent_trainer_and_dataloaders_deepcopy(self) -> None:
+        self._prevent_trainer_and_dataloaders_deepcopy = True
+        yield
+        self._prevent_trainer_and_dataloaders_deepcopy = False
+
+    def __getstate__(self) -> Dict[str, Any]:
+        skip_keys = set()
+        if self._prevent_trainer_and_dataloaders_deepcopy:
+            skip_keys.update({
+                "trainer", "train_dataloader", "val_dataloader", "test_dataloader", "predict_dataloader", "datamodule"
+            })
+        return {k: v for k, v in self.__dict__.items() if k not in skip_keys}
```

Comment:
I prefer the implementation where the skip_keys are set to None. It's weird when the property doesn't exist but it's defined in the source code

Reply:
We should skip it as setting None would be override the base function if they exist. This would be wrong.